### PR TITLE
Ygg-api support

### PR DESCRIPTION
**Pull request by Clemv95**
#### Description
Add support for ygg-api
Alternative way of getting torrents from ygg
it doesnt need cloudflare bypass
https://yggapi.eu/

### CODE REVIEW

**Review by garfield69 — CHANGES_REQUESTED** [review]
failing lint
```
  1:1       error    missing document start "---"  (document-start)
  109:1     error    too many blank lines (2 > 1)  (empty-lines)
  121:1     error    trailing spaces  (trailing-spaces)
```

File: src/Jackett.Common/Definitions/ygg-api.yml [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+          args: \b(2160p|1080p|720p|480p)\b
+    seeders:
+      selector: seeders
+    leechers:
+      selector: leechers
+    date:
+      selector: uploaded_at
+      filters:
+        - name: dateparse
+          args: 2006-01-02T15:04:05-07:00
```
Comment (garfield69) [inline src/Jackett.Common/Definitions/ygg-api.yml:180]: https://github.com/Jackett/Jackett/wiki/Definition-format#dateparse

File: src/Jackett.Common/Definitions/ygg-api.yml [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+id: yggapi
+name: Ygg API
+description: Indexeur non officiel pour ygg.re
+language: fr-FR
+type: private
+encoding: UTF-8
+followredirect: true
+testlinktorrent: false
+links:
+  - https://yggapi.eu
```
Comment (garfield69) [inline src/Jackett.Common/Definitions/ygg-api.yml:11]:
```suggestion
  - https://yggapi.eu/
```
standard for links is to end with a slash. don't forget to remove the slash at the start of all the search paths ;-)

File: src/Jackett.Common/Definitions/ygg-api.yml [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    download:
+      selector: id
+      filters:
+        - name: prepend
+          args: https://yggapi.eu/torrent/
+        - name: append
+          args: /download?passkey={{ .Config.passkey }}
+    size:
+      selector: size
+    _quality:
```
Comment (garfield69) [inline src/Jackett.Common/Definitions/ygg-api.yml:167]: quality is not a standard field, in that it is not returned in torznab results.
so it is not necessary to extract it unless you plan to incorporate it into one of the other fields,
such as description for example. or you could just drop it altogether.

**Review by garfield69 — CHANGES_REQUESTED** [review]

File: src/Jackett.Common/Definitions/ygg-api.yml [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+      selector: leechers
+    date:
+      selector: uploaded_at
+      filters:
+        - name: dateparse
+          args: "yyyy-MMM-dd HH:mm:sszzz"
+          
+          
+    downloadvolumefactor:
+      text: "1"
```
Comment (garfield69) [inline src/Jackett.Common/Definitions/ygg-api.yml:164]:
```suggestion
      text: 1
```

File: src/Jackett.Common/Definitions/ygg-api.yml [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    uploadvolumefactor:
+      text: "1"
```
Comment (garfield69) [inline src/Jackett.Common/Definitions/ygg-api.yml:166]:
```suggestion
      text: 1
# JSON 1.0.0
```

**Review by garfield69 — CHANGES_REQUESTED** [review]
we usually add an engine tag at the bottom ;-)

File: src/Jackett.Common/Definitions/ygg-api.yml (hunk 2 above)
Comment (Clemv95) [reply src/Jackett.Common/Definitions/ygg-api.yml]: Done ;)

**Comment by Clemv95** [comment]
This is my first contribution, sorry for all the few mistakes, especially compliance one

**Comment by garfield69** [comment]
no problem, it is a process of improving bit by bit, especially given that Jackett's implementation of yaml is quirky at best.
there is a full wiki on yaml definitions available, at https://github.com/Jackett/Jackett/wiki/Definition-format but it is quite a info overload for first time reading ;-)

and the pipeline error traps are not easy to decipher, so that can complicate resolution.

at some point I will download and test the indexer and see what other updates may be needed.

**Comment by Clemv95** [comment]
> no problem, it is a process of improving bit by bit, especially given that Jackett's implementation of yaml is quirky at best. there is a full wiki on yaml definitions available, at https://github.com/Jackett/Jackett/wiki/Definition-format but it is quite a info overload for first time reading ;-)
> 
> and the pipeline error traps are not easy to decipher, so that can complicate resolution.
> 
> at some point I will download and test the indexer and see what other updates may be needed.

Thanks !
I tried ajv but i struggle to understand what the errors means and how to fix them

**Comment by garfield69** [comment]
Ok so we finally passed all the checks, so please do not make any further changes ATM.
I will now copy the indexer, update it to conform to our common standards and test it till I'm happy its fit for purpose ;-)
BTW, thanks for contributing!

**Review by garfield69 — APPROVED** [review]

**Comment by garfield69** [comment]
I added some config info, fix date processing, simplified the path because there was no special processing on any of the categories, and brought it in line with our usual json indexer layouts.

if you have no objections let me know, otherwise I will merge later today to be included with the next Jackett release due out in about 7 hours from now.

**Comment by Clemv95** [comment]
> I added some config info, fix date processing, simplified the path because there was no special processing on any of the categories, and brought it in line with our usual json indexer layouts.
> 
> if you have any objections let me know, otherwise I will merge later today to be included with the next Jackett release due out in about 7 hours from now.

I will check that in few hours and let you know

**Comment by Clemv95** [comment]
> > I added some config info, fix date processing, simplified the path because there was no special processing on any of the categories, and brought it in line with our usual json indexer layouts.
> > if you have any objections let me know, otherwise I will merge later today to be included with the next Jackett release due out in about 7 hours from now.
> 
> I will check that in few hours and let you know

I have tested on my side, and its ok !

**Comment by garfield69** [comment]
I have reverted the merging of this PR.
I have been thinking about this and I realized that what this amounts to is bypassing the YGG users account download quota.
This is not acceptable. We will not support any method that effectively provides torrents from a private site without permission.

**Comment by Clemv95** [comment]
> I have reverted the merging of this PR. I have been thinking about this and I realized that what this amounts to is bypassing the YGG users account download quota. This is not acceptable. We will not support any method that effectively provides torrents from a private site without permission.

It doesnt bypass download quota because it use passkey to download torrents
It steel update download ratio i can confirm

**Comment by garfield69** [comment]
So if a user is blocked from downloading a torrent on the ygg website because their quota is exceeded, you are saying that if they use the ygg-api to initiate a download, it will fail for them?

**Comment by Clemv95** [comment]
> So if a user is blocked from downloading a torrent on the ygg website because their quota is exceeded, you are saying that if they use the ygg-api to initiate a download, it will fail for them?

I will try
But i think there is two solutions :
First one the download from api succeed, but once in a download client it fail to download
Second one the download fail from the api

**Comment by Clemv95** [comment]
> > So if a user is blocked from downloading a torrent on the ygg website because their quota is exceeded, you are saying that if they use the ygg-api to initiate a download, it will fail for them?
> 
> I will try But i think there is two solutions : First one the download from api succeed, but once in a download client it fail to download Second one the download fail from the api

Okay this is the first option 
The download succeeded but if the ratio is negative it will not download torrent data from client 
But we can still download .torrent file 
Dont know if you see an issue about being able to download .torrent but not the data

**Comment by Clemv95** [comment]
> > > So if a user is blocked from downloading a torrent on the ygg website because their quota is exceeded, you are saying that if they use the ygg-api to initiate a download, it will fail for them?
> > 
> > 
> > I will try But i think there is two solutions : First one the download from api succeed, but once in a download client it fail to download Second one the download fail from the api
> 
> Okay this is the first option The download succeeded but if the ratio is negative it will not download torrent data from client But we can still download .torrent file Dont know if you see an issue about being able to download .torrent but not the data

The only downside i can see is that it is possible to download .torrent from ygg without having an account by the ygg-api
But without a valid passkey the download will fail so...

**Comment by Clemv95** [comment]
> > > > So if a user is blocked from downloading a torrent on the ygg website because their quota is exceeded, you are saying that if they use the ygg-api to initiate a download, it will fail for them?
> > > 
> > > 
> > > I will try But i think there is two solutions : First one the download from api succeed, but once in a download client it fail to download Second one the download fail from the api
> > 
> > 
> > Okay this is the first option The download succeeded but if the ratio is negative it will not download torrent data from client But we can still download .torrent file Dont know if you see an issue about being able to download .torrent but not the data
> 
> The only downside i can see is that it is possible to download .torrent from ygg without having an account by the ygg-api But without a valid passkey ( link to an account with a positive ratio) the download will fail so...

The passkey is a parameter of the ygg tracker, which is mandatory to download
https://yggland.fr/FAQ-Tutos/#accordion-tab2-content1

**Comment by garfield69** [comment]
I need feedback from the other team members before I continue.
@ilike2burnthing  @mynameisbogdan  opinion requested: if an unofficial api is able to provide a torrent meta file from a private site using a users' passkey (provided by the user with an account on ygg), even though if they had used the ygg website they would not have been able to fetch it because the web site would have blocked the download, would that constitute a rule violation?
Having fetched the meta from the api, and loaded it on a torrent client, the ygg tracker would apparently refuse the connect by the client to the tracker and allow peer exchange to download data.
Should we implement this indexer?

**Comment by Clemv95** [comment]
> I need feedback from the other team members before I continue. @ilike2burnthing @mynameisbogdan opinion requested: if an unofficial api is able to provide a torrent meta file from a private site using a users' passkey (provided by the user with an account on ygg), even though if they had used the ygg website they would not have been able to fetch it because the web site would have blocked the download, would that constitute a rule violation? Having fetched the meta from the api, and loaded it on a torrent client, the ygg tracker would apparently refuse the connect by the client to the tracker and [LE] **not** allow peer exchange to download data. Should we implement this indexer?

Just a precision :
On ygg website, if our ratio is negative, we can still download .torrent file, and after same behavior, impossible to download from torrent client
This is to allow users to seed torrents to raise their ratio and not be negative

**Comment by Clemv95** [comment]
> I don't know about the torrent files, since even if users should be able to grab them the tracker should prevent data downloading in the dl client.
> 
> I'm more against this unofficial API due to being a MITM and can do anything with the credentials, and due to previous interactions with ygg users most likely we'll get the blame is something wrong happens.
> 
> As such, we won't implement this on Prowlarr side.

I can understand
Maybe just put a disclaimer then, users who use this api are aware of this possibility, and if something happen, they can just blame themself, its a risk

**Comment by garfield69** [comment]
>I'm more against this unofficial API due to being a MITM and can do anything with the credentials

Yes, valid point. Especially since I cannot find in YGG any use option to change the passkey, like you can on other private sites.

Possible circumvention: What about replacing the download link that uses the yggapi domain, and using the official download link from YGG site.
So we add the regular username/password login from the ygg indexer to the yggapi indexer, and essentially we use the yggapi only for searching, but downloading is done from YGG?

**Comment by Clemv95** [comment]
> > I'm more against this unofficial API due to being a MITM and can do anything with the credentials
> 
> Yes, valid point. Especially since I cannot find in YGG any user option to change the passkey, like you can on other private sites.
> 
> Possible circumvention: What about replacing the download link that uses the yggapi domain, and using the official download link from YGG site. So we add the regular username/password login from the ygg indexer to the yggapi indexer, and essentially we use the yggapi only for searching, but downloading is done from YGG?

This could be a nice solution, but this will implicated the need of flaresolverr 
And the Idea of ygg-api is to bypass this need which is annoying

**Comment by Clemv95** [comment]
> > Possible circumvention: What about replacing the download link that uses the yggapi domain, and using the official download link from YGG site.
> 
> Are the YGG download links stateless with some auth key, or do they require cookies from an authenticated user? If they aren't, we can't auth on two sites in the same time.
You need to be authenticated to download torrent file from ygg

**Comment by garfield69** [comment]
As a POC, I set up on my native windows 10 Jackett a version of the ygg-api that uses the ygg username/password login and the ygg download path.
And with the Jackett dashboard did a keywordless search to list some torrents and clicked on the download link for a couple of them, and they both successfully started the download prompt for saving the .torrent file to disk.
I cancelled the save because I do not want to actually download anything, but I've no reason to believe the meta file would not be good for loading to a client.
However, as pointed out by the OP earlier, the login is prefixed by a cloudlfare check which according to the OP defeats the main objective of using this unofficial api service in the first place.
I run win10 desktop docker, with the 21hsmw/flaresolverr:nodriver image which has been solving all my cloudflare challenges which is how I am able to use all my indexers that have challengers ATM.

I'm still hoping to get feedback from the other team member, but at this stage I don't think we will be including either version of the ygg-api indexer to the Jackett stable.

**Comment by garfield69** [comment]
Oh, an of course the MITM exposure is still present even when using the user/pswd method, because the api server would be able to see the cookies that contain the authorisation for the ygg login if it chose to look at them.

**Comment by Clemv95** [comment]
Any update ?

**Comment by garfield69** [comment]
Due to the MITM exposure we will not be officially adding this indexer to Jackett.

**Comment by ilike2burnthing** [comment]
I'm very late to game here, but just to add something that may be of some use - the indexers for TorrentDay and IPTorrents include unofficial mirrors, and to my knowledge they're the only semi-/private indexers which do, however if I'm remembering correctly they're mirrors hosted by trusted members of the sites, and have been okayed (even promoted) by site staff.

Basically, that's the standard you'd need to meet for us to entertain this - if YGG staff make a post we can see that says it's cool, then we're cool.

**Pull request merged**